src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,289 @@
+            }
+
+            if ((HttpLoggingFields.Response & _options.LoggingFields) == HttpLoggingFields.None)
+            {
+                // Short circuit and don't replace response body.
+                await _next(context).ConfigureAwait(false);
+                return;
+            }
+
+            var response = context.Response;
+
+            ResponseBufferingStream? bufferingStream = null;
+            IHttpResponseBodyFeature? originalBodyFeature = null;
+            if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody))
+            {
+                originalBodyFeature = context.Features.Get<IHttpResponseBodyFeature>()!;
+                // TODO pool these.
+                bufferingStream = new ResponseBufferingStream(originalBodyFeature, _options.ResponseBodyLogLimit);
+                response.Body = bufferingStream;
+            }
+
+            try
+            {
+                await _next(context).ConfigureAwait(false);
+                var list = new List<KeyValuePair<string, object?>>();
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.StatusCode))
+                {
+                    list.Add(new KeyValuePair<string, object?>(nameof(response.StatusCode), response.StatusCode));
+                }
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseHeaders))
+                {
+                    FilterHeaders(list, response.Headers, _options.AllowedResponseHeaders);
+                }
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody) && IsSupportedMediaType(response.ContentType))
+                {
+                    var body = bufferingStream!.GetString(_options.BodyEncoding);
+                    list.Add(new KeyValuePair<string, object?>(nameof(response.Body), body));
+                }
+
+                var httpResponseLog = new HttpResponseLog(list);
+
+                _logger.Log(LogLevel.Information,
+                    eventId: LoggerEventIds.ResponseLog,
+                    state: httpResponseLog,
+                    exception: null,
+                    formatter: HttpResponseLog.Callback);
+            }
+            finally
+            {
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody))
+                {
+                    if (bufferingStream != null)
+                    {
+                        bufferingStream.Dispose();
+                    }
+
+                    context.Features.Set(originalBodyFeature);
+                }
+            }
+        }
+
+        private bool IsSupportedMediaType(string contentType)
+        {
+            var mediaTypeList = _options.SupportedMediaTypes;
+            if (mediaTypeList == null || mediaTypeList.Count == 0 || string.IsNullOrEmpty(contentType))
+            {
+                return false;
+            }
+
+            var mediaType = new MediaTypeHeaderValue(contentType);
```

Comment:
This allocates...

Reply:
Yep, I could use the MVC MediaType value here as it's a struct. Not sure what alternatives I have here though if we don't want to use that type.